src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
-	return 0;
+		mem_handle = MEM_GetNextFreePage();
+		if (!mem_handle) {
+			// Windows 3.1 does this really often
+			// LOG_MSG("XMS: Allocate zero pages with no memory left");
+		}
+	}
+
+	xms.handles[index].is_free    = false;
+	xms.handles[index].mem_handle = mem_handle;
+	xms.handles[index].lock_count = 0;
+	xms.handles[index].size_kb    = size_kb;
+
+	handle = index;
+	return Result::OK;
 }
 
-Bitu XMS_FreeMemory(Bitu handle) {
-	if (InvalidHandle(handle)) return XMS_INVALID_HANDLE;
-	MEM_ReleasePages(xms_handles[handle].mem);
-	xms_handles[handle].mem=-1;
-	xms_handles[handle].size=0;
-	xms_handles[handle].free=true;
-	return 0;
+static Result xms_free_memory(const uint16_t handle)
+{
+	if (!xms_is_handle_valid(handle)) {
+		return Result::XMS_INVALID_HANDLE;
+	}
+	if (xms.handles[handle].lock_count != 0) {
+		return Result::XMS_BLOCK_LOCKED;
+	}
+
+	MEM_ReleasePages(xms.handles[handle].mem_handle);
+
+	xms.handles[handle] = XMS_Block();
+	return Result::OK;
 }
 
-Bitu XMS_MoveMemory(PhysPt bpt) {
-	/* Read the block with mem_read's */
-	Bitu length=mem_readd(bpt+offsetof(XMS_MemMove,length));
-	Bitu src_handle=mem_readw(bpt+offsetof(XMS_MemMove,src_handle));
+static Result xms_move_memory(const PhysPt bpt)
+{
+	// TODO: Detect invalid overlaps, report XMS_INVALID_OVERLAP
+
+	// Read the block with mem_read's
+	const auto length = mem_readd(bpt + offsetof(XMS_MemMove, length));
+
+	// "Length must be even" --Microsoft XMS Spec 3.0
+	if (length % 2) {
+		return Result::XMS_PARITY_ERROR;
+	}
+
 	union {
 		RealPt realpt;
 		uint32_t offset;
-	} src,dest;
-	src.offset=mem_readd(bpt+offsetof(XMS_MemMove,src.offset));
-	Bitu dest_handle=mem_readw(bpt+offsetof(XMS_MemMove,dest_handle));
-	dest.offset=mem_readd(bpt+offsetof(XMS_MemMove,dest.offset));
-	PhysPt srcpt,destpt;
+	} src, dst;
+
+	const auto src_handle = mem_readw(
+		static_cast<PhysPt>(bpt + offsetof(XMS_MemMove, src_handle)));
+	src.offset = mem_readd(
+		static_cast<PhysPt>(bpt + offsetof(XMS_MemMove, src.offset)));
+
+	const auto dst_handle = mem_readw(
+		static_cast<PhysPt>(bpt + offsetof(XMS_MemMove, dst_handle)));
+	dst.offset = mem_readd(
+		static_cast<PhysPt>(bpt + offsetof(XMS_MemMove, dst.offset)));
+
+	PhysPt srcpt = 0;
+	PhysPt dstpt = 0;
+
 	if (src_handle) {
-		if (InvalidHandle(src_handle)) {
-			return XMS_INVALID_SOURCE_HANDLE;
+		if (!xms_is_handle_valid(src_handle)) {
+			return Result::XMS_INVALID_SRC_HANDLE;
 		}
-		if (src.offset>=(xms_handles[src_handle].size*1024U)) {
-			return XMS_INVALID_SOURCE_OFFSET;
+		if (src.offset >= (xms.handles[src_handle].size_kb * 1024U)) {
+			return Result::XMS_INVALID_SRC_OFFSET;
 		}
-		if (length>xms_handles[src_handle].size*1024U-src.offset) {
-			return XMS_INVALID_LENGTH;
+		if (length > xms.handles[src_handle].size_kb * 1024U - src.offset) {
+			return Result::XMS_INVALID_LENGTH;
 		}
-		srcpt=(xms_handles[src_handle].mem*4096)+src.offset;
+		srcpt = (static_cast<uint32_t>(xms.handles[src_handle].mem_handle) * 4096U) +
```

Comment:
Introduce page_size_bytes for 4096 or something.

Reply:
Done.